File: src/com/eagle/programmar/Javascript/Statements/Javascript_DocumentWriteln.java
// Copyright Eagle Legacy Modernization, 2010-date
// Original author: Steven A. O'Hara, Apr 1, 2024

package com.eagle.programmar.Javascript.Statements;

import com.eagle.interpret.EagleInterpreter;
import com.eagle.interpret.EagleRunnable;
import com.eagle.programmar.Javascript.Javascript_Expression;
import com.eagle.programmar.Javascript.Terminals.Javascript_Keyword;
import com.eagle.programmar.Javascript.Terminals.Javascript_KeywordChoice;
import com.eagle.tokens.PrimaryOperator;
import com.eagle.tokens.interfaces.AbstractExpression;
import com.eagle.tokens.interfaces.AbstractStatement;
import com.eagle.tokens.punctuation.PunctuationLeftParen;
import com.eagle.tokens.punctuation.PunctuationPeriod;
import com.eagle.tokens.punctuation.PunctuationRightParen;
import com.eagle.tokens.punctuation.PunctuationSemicolon;
import com.eagle.transform.EagleGenerator;
import com.eagle.transform.EagleTransformableStatement;
import com.eagle.transform.EagleTransformer;

public class Javascript_DocumentWriteln extends PrimaryOperator
		implements EagleRunnable, EagleTransformableStatement
{
	public @S(10) Javascript_Keyword DOCUMENT = new Javascript_Keyword("document");
	public @S(20) PunctuationPeriod dot;
	public @S(30) Javascript_KeywordChoice WRITELN = new Javascript_KeywordChoice(
			"write", "writeln");
	public @S(40) PunctuationLeftParen leftParen;
	public @S(50) Javascript_Expression expr;
	public @S(60) PunctuationRightParen rightParen;
	public @S(70) PunctuationSemicolon semicolon;

	@Override
	public void interpret(EagleInterpreter interpreter)
	{
		boolean newLine;
		switch (WRITELN.getValue())
		{
		case "write":
			newLine = false;
			break;
		case "writeln":
			newLine = true;
			break;
		default:
			throw new RuntimeException("Unexpected WRITELN value: " + WRITELN.getValue());
		}

		String val = interpreter.getStrValue(expr);
		if (val.startsWith("<br>")) val = val.substring(4);	// Toss leading <br> if present
		if (newLine)
		{
			System.out.println(val);
		}
		else
		{
			System.out.print(val);
		}
	}

	@Override
	public AbstractStatement transformStatement(EagleTransformer transformer,
			EagleGenerator generator)
	{
		boolean newLine;
		switch (WRITELN.getValue())
		{
		case "write":
			newLine = false;
			break;
		case "writeln":
			newLine = true;
			break;
		default:
			throw new RuntimeException("Unexpected WRITE value: " + WRITELN.getValue());
		}
		
		AbstractExpression value = transformer.transformExpression(generator, expr);
		return generator.newPrintStatement(value, newLine, false, this);
	}
}

File: csrc/com/eagle/programmar/Javascript/Statements/Javascript_DocumentWriteln.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// Purchase a Premium Edition license at:
// https://www.tangiblesoftwaresolutions.com/order/order-java-to-csharp.html
// ====================================================================================================

using System;

// Copyright Eagle Legacy Modernization, 2010-date
// Original author: Steven A. O'Hara, Apr 1, 2024

namespace com.eagle.programmar.Javascript.Statements
{
	using EagleInterpreter = com.eagle.interpret.EagleInterpreter;
	using EagleRunnable = com.eagle.interpret.EagleRunnable;
	using Javascript_Expression = com.eagle.programmar.Javascript.Javascript_Expression;
	using Javascript_Keyword = com.eagle.programmar.Javascript.Terminals.Javascript_Keyword;
	using Javascript_KeywordChoice = com.eagle.programmar.Javascript.Terminals.Javascript_KeywordChoice;
	using PrimaryOperator = com.eagle.tokens.PrimaryOperator;
	using AbstractExpression = com.eagle.tokens.interfaces.AbstractExpression;
	using AbstractStatement = com.eagle.tokens.interfaces.AbstractStatement;
	using AbstractType = com.eagle.tokens.interfaces.AbstractType;
	using AbstractVariable = com.eagle.tokens.interfaces.AbstractVariable;
	using PunctuationLeftParen = com.eagle.tokens.punctuation.PunctuationLeftParen;
	using PunctuationPeriod = com.eagle.tokens.punctuation.PunctuationPeriod;
	using PunctuationRightParen = com.eagle.tokens.punctuation.PunctuationRightParen;
	using PunctuationSemicolon = com.eagle.tokens.punctuation.PunctuationSemicolon;
	using EagleGenerator = com.eagle.transform.EagleGenerator;
	using EagleTransformableStatement = com.eagle.transform.EagleTransformableStatement;
	using EagleTransformer = com.eagle.transform.EagleTransformer;
	using TypeEnum = com.eagle.transform.EagleGenerator.TypeEnum;

	public class Javascript_DocumentWriteln : PrimaryOperator, EagleRunnable, EagleTransformableStatement
	{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.programmar.Javascript.Terminals.Javascript_Keyword DOCUMENT = new com.eagle.programmar.Javascript.Terminals.Javascript_Keyword("document");
		public Javascript_Keyword DOCUMENT = new Javascript_Keyword("document");
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.tokens.punctuation.PunctuationPeriod dot;
		public PunctuationPeriod dot;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(30) com.eagle.programmar.Javascript.Terminals.Javascript_KeywordChoice WRITELN = new com.eagle.programmar.Javascript.Terminals.Javascript_KeywordChoice("write", "writeln");
		public Javascript_KeywordChoice WRITELN = new Javascript_KeywordChoice("write", "writeln");
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(40) com.eagle.tokens.punctuation.PunctuationLeftParen leftParen;
		public PunctuationLeftParen leftParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(50) com.eagle.programmar.Javascript.Javascript_Expression expr;
		public Javascript_Expression expr;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(60) com.eagle.tokens.punctuation.PunctuationRightParen rightParen;
		public PunctuationRightParen rightParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(70) com.eagle.tokens.punctuation.PunctuationSemicolon semicolon;
		public PunctuationSemicolon semicolon;

		public override void interpret(EagleInterpreter interpreter)
		{
			bool newLine;
			switch (WRITELN.getValue())
			{
			case "write":
				newLine = false;
				break;
			case "writeln":
				newLine = true;
				break;
			default:
				throw new Exception("Unexpected WRITELN value: " + WRITELN.getValue());
			}

			string val = interpreter.getStrValue(expr);
			if (val.StartsWith("<br>", StringComparison.Ordinal))
			{
				val = val.Substring(4); // Toss leading <br> if present
			}
			if (newLine)
			{
				Console.WriteLine(val);
			}
			else
			{
				Console.Write(val);
			}
		}

		public override AbstractStatement transformStatement(EagleTransformer transformer, EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
		{
			bool newLine;
			switch (WRITELN.getValue())
			{
			case "write":
				newLine = false;
				break;
			case "writeln":
				newLine = true;
				break;
			default:
				throw new Exception("Unexpected WRITE value: " + WRITELN.getValue());
			}

			AbstractExpression value = transformer.transformExpression(generator, expr);
			return generator.newPrintStatement(value, EagleGenerator.TypeEnum.STRING, newLine, false, this);
		}
	}

}
